src/contrib/google/auth.ts

```
@@ -119,3 +126,60 @@ export async function handleGoogleRequestRejection(
 
   return new Error(getErrorMessage(error.result.error));
 }
+
+export async function handleGoogleRequestRejection(
+  error: unknown,
+  googleAccount: SanitizedIntegrationConfig | null,
+  legacyClientToken: string | null
+): Promise<Error> {
+  const basicMessage = "Google rejected request";
+  console.debug(basicMessage, { error });
+
+  if (!isObject(error)) {
+    return new Error(basicMessage, { cause: error });
+  }
+
+  if (!isAxiosError(error) || error.response == null) {
+    return {
+      name: "UnknownError",
+      message: basicMessage,
+      ...error,
+    };
+  }
+
+  const { status } = error.response;
+
+  if (status === 404) {
+    return new PermissionsError(
+      "Cannot locate the Google drive resource. Have you been granted access?",
+      status
+    );
+  }
+
+  if ([403, 401].includes(status)) {
+    if (legacyClientToken) {
+      await chromeP.identity.removeCachedAuthToken({
+        token: legacyClientToken,
+      });
+      console.debug(
+        "Bad Google (legacy) OAuth token. Removed the auth token from the cache so the user can re-authenticate"
+      );
+    } else if (googleAccount) {
+      await deleteCachedAuthData(googleAccount.id);
+      console.debug(
+        "Bad Google client PKCE token. Removed the auth token from the cache so the user can re-authenticate"
+      );
+    } else {
+      console.debug("No auth token provided for request");
+    }
+
+    return new PermissionsError(
+      `Permission denied, re-authenticate with Google and try again. Details: ${getErrorMessage(
+        error
+      )}`,
+      status
+    );
+  }
+
+  return new Error(getErrorMessage(error));
```

Why not just return the original error here?

---

Typing, mostly. Would have to hard-cast-plus-@ts-ignore on the `unknown` input `error`

---

Oh looks like lint/TS actually don't complain about a regular cast 🤷‍♂️